<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="description" content="None" />
      <link rel="shortcut icon" href="img/favicon.ico" />
    <title>trusTEr</title>
    <link rel="stylesheet" href="css/theme.css" />
    <link rel="stylesheet" href="css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Introduction";
        var mkdocs_page_input_path = "index.md";
        var mkdocs_page_url = null;
      </script>
    
    <script src="js/jquery-3.6.0.min.js" defer></script>
    <!--[if lt IE 9]>
      <script src="js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/highlight.min.js"></script>
      <script>hljs.initHighlightingOnLoad();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="." class="icon icon-home"> trusTEr
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="./search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
                <li class="toctree-l1 current"><a class="reference internal current" href=".">Introduction</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#requirements">Requirements</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#how-to-install">How to install</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#just-the-modules">Just the modules</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#with-docker-container">With Docker container</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#with-conda-environment">With conda environment</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#introduction">Introduction</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#structure">Structure</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#functionality-and-workflow">Functionality and workflow</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#running-on-a-server">Running on a server</a>
    </li>
    </ul>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="workflow/">Workflow</a>
                </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href=".">trusTEr</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="." class="icon icon-home" alt="Docs"></a> &raquo;</li>
      <li>Introduction</li>
    <li class="wy-breadcrumbs-aside">
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="truster-a-trusting-te-cluster-analysis">trusTEr - A trusting TE cluster analysis</h1>
<p><strong>Version 0.1.1, written in Python 3.6.6</strong></p>
<p>Takes fastq files from 10x single cell RNA sequencing, clusters cells using Seurat, and can be used to produce 
read count matrices in a cluster level. You can also quantify reads per cluster having predefined clusters.</p>
<h2 id="requirements">Requirements</h2>
<p>TrusTEr depends on several external software. We provide a Docker container and a conda environment for a quick-start. </p>
<p>TrusTEr requires:</p>
<ul>
<li>Cellranger</li>
<li>R (version 3.6)<ul>
<li>Seurat</li>
</ul>
</li>
<li>TEtranscripts</li>
<li>STAR aligner</li>
<li>subset-bam and bamtofastq from 10x Genomics</li>
<li>Velocyto</li>
</ul>
<p>The package has been tested in Unix systems only and supports only SLURM job submissions.</p>
<h2 id="how-to-install">How to install</h2>
<h4 id="just-the-modules">Just the modules</h4>
<p>If you fulfill the requirements, you can install via pip:
<code>pip install truster</code></p>
<h4 id="with-docker-container">With Docker container</h4>
<h4 id="with-conda-environment">With conda environment</h4>
<h2 id="introduction">Introduction</h2>
<p>As single cell technologies haven't developed to the point where we can get the needed sequencing depth to study transposable elements expression, trusTEr seeks produce more reliable results by combining reads from closely related cells to gain information on a cell type level.</p>
<p><img src="https://raw.githubusercontent.com/ra7555ga-s/truster/main/img/approach.png" width="700"></p>
<h2 id="structure">Structure</h2>
<p>trusTEr uses composition assiciation to relate three main classes: </p>
<ul>
<li>Experiment: Includes information about the experiment as a whole. This is the main object you will be working with.<ul>
<li>Name is required, description is optional. </li>
<li>Register samples by providing a path to its fastq files.</li>
</ul>
</li>
<li>Sample: Created by giving a path to fastq files<ul>
<li>Name and ID required. </li>
</ul>
</li>
<li>Cluster: Created by running <code>get_clusters()</code> or <code>merge_samples()</code> functions (Or <code>set_clusters_outdir()</code> or <code>set_merge_samples_outdir()</code>).</li>
</ul>
<p><img src="https://raw.githubusercontent.com/ra7555ga-s/truster/main/img/compositionAssociation.png" width="500"></p>
<h2 id="functionality-and-workflow">Functionality and workflow</h2>
<p>This package is meant to be run with the following workflow:</p>
<p><img src="https://raw.githubusercontent.com/ra7555ga-s/truster/main/img/workflow.png" width="470"></p>
<p>Depending on the object type, you have access to different functions to go through these steps. </p>
<p><code>Experiment</code> is the main object one would work with. Here you will <code>register_sample()</code> or <code>register_samples_from_path()</code>. </p>
<p>An object of type <code>Sample</code> has access to step 7 and some handy wrappers to use <code>cellRanger</code>, perform clustering with Seurat and run and plot RNA velocity. These functions can be called for all registered samples from your object of type <code>Experiment</code> (See <code>quantify()</code>, <code>get_clusters_all_samples()</code>, <code>velocity_all_samples()</code>, <code>plot_velocity_all_samples()</code>).</p>
<p>The need for the class <code>Experiment</code> is clearer once the user wants to merge samples (See <code>merge_samples()</code>) or to run the same workflow for all the samples' clusters.</p>
<p>The user won't work directly with an object of type <code>Cluster</code>, but this class includes all the functions needed to go through steps 1-6 of the workflow. Instead of running this pipeline individually for each cluster, one can run the workflow for each cluster of each registered sample using the <code>Experiment</code> function <code>process_clusters(mode = "per_sample", ...)</code>.</p>
<p>One can also partition the workflow and run step by step in all registered samples or in a combined clustering using the transitioning functions of <code>Experiment</code> to call the needed functions in the class <code>Cluster</code> (See <code>tsv_to_bam_clusters()</code>, <code>filter_UMIs_clusters()</code>, <code>bam_to_fastq_clusters()</code>, <code>concatenate_lanes_clusters()</code>, <code>merge_clusters()</code>, <code>map_clusters()</code>, <code>TE_counts_clusters()</code>, <code>normalize_TE_counts()</code>).</p>
<h2 id="running-on-a-server">Running on a server</h2>
<p>At the moment of construction of an <code>Experiment</code> object, you can declare <code>slurm_path</code> and a <code>modules_path</code>. These files will be checked whenever a function can be run as an <code>sbatch</code> job.</p>
<p>For the moment, trusTEr only works with <code>slurm</code> systems. You need to create two json files:</p>
<p><strong>1. slurmPath</strong>
Declaring the <a href="https://slurm.schedmd.com/sbatch.html">sbatch options</a> per function. You need to declare a field per function listed that you will use:</p>
<ul>
<li>quantify</li>
<li>get_clusters</li>
<li>merge_samples</li>
<li>tsv_to_bam</li>
<li>filter_UMIs</li>
<li>bam_to_fastq</li>
<li>map_cluster</li>
<li>TE_count</li>
<li>velocity</li>
<li>plot_velocity</li>
</ul>
<p>As a brief example:</p>
<pre><code>{
  &quot;__default__&quot; :
  {
      &quot;account&quot; : &quot;myaccount&quot;,
      &quot;time&quot; : &quot;00:15:00&quot;,
      &quot;nodes&quot; : 1,
      &quot;partition&quot; : &quot;dell&quot;,
      &quot;output&quot; : &quot;%j.out&quot;,
      &quot;error&quot; : &quot;%j.err&quot;,
      &quot;job-name&quot; : &quot;%j&quot;
  },
  &quot;quantify&quot; :
  {
      &quot;account&quot; : &quot;myaccount&quot;,
      &quot;time&quot; : &quot;10:00:00&quot;,
      &quot;nodes&quot; : 1,
      &quot;tasks-per-node&quot; : 20,
      &quot;partition&quot; : &quot;dell&quot;,
      &quot;output&quot; : &quot;%j.quantify.out&quot;,
      &quot;error&quot; : &quot;%j.quantify.err&quot;,
      &quot;job-name&quot; : &quot;%j.quantify&quot;
  },
  &quot;get_clusters&quot; : {...},
  &quot;merge_samples&quot; : {...},
  &quot;tsv_to_bam&quot; : {...},  
  &quot;filter_UMIs&quot; : {...},
  &quot;bam_to_fastq&quot; : {...},
  &quot;map_cluster&quot; : {...},
  &quot;TE_count&quot; : {...},
  &quot;velocity&quot; : {...},
  &quot;plot_velocity&quot; : {...}
}
</code></pre>
<p><strong>2. modulesPath</strong></p>
<p>In a <code>slurm</code> system, many times you need to load modules for a software to be available. In this json file you declare the name of the needed modules. You need to declare a field per function listed that you will use:</p>
<ul>
<li>quantify</li>
<li>get_clusters</li>
<li>merge_samples</li>
<li>tsv_to_bam</li>
<li>filter_UMIs</li>
<li>bam_to_fastq</li>
<li>map_cluster</li>
<li>TE_count</li>
<li>TE_count_unique</li>
<li>velocity</li>
<li>normalize_TE_counts</li>
<li>plot_TE_expression</li>
</ul>
<p>As an example:</p>
<pre><code>{
        &quot;quantify&quot;:[&quot;cellranger/3.1.0&quot;],
        &quot;velocity&quot;:[&quot;GCC/7.3.0-2.30&quot;, &quot;SAMtools/1.9&quot;, &quot;velocyto/0.17.17&quot;],
        &quot;get_clusters&quot;:[&quot;GCC/9.3.0&quot;, &quot;OpenMPI/4.0.3&quot;, &quot;Seurat/3.1.5-R-4.0.0&quot;],
        &quot;merge_samples&quot;:[&quot;GCC/9.3.0&quot;, &quot;OpenMPI/4.0.3&quot;, &quot;Seurat/3.1.5-R-4.0.0&quot;],
        &quot;tsv_to_bam&quot;:[&quot;subset-bam/1.0&quot;],
        &quot;filter_UMIs&quot;:[&quot;GCC/7.3.0-2.30&quot;,  &quot;OpenMPI/3.1.1&quot;, &quot;Pysam/0.15.1-Python-3.6.6&quot;],
        &quot;bam_to_fastq&quot;:[&quot;bamtofastq/1.2.0&quot;],
        &quot;map_cluster&quot;:[&quot;GCC/5.4.0-2.26&quot;, &quot;OpenMPI/1.10.3&quot;, &quot;STAR/2.6.0c&quot;],
        &quot;TE_count&quot;:[&quot;icc/2018.1.163-GCC-6.4.0-2.28&quot;, &quot;OpenMPI/2.1.2&quot;, &quot;TEToolkit/2.0.3-Python-2.7.14&quot;],
        &quot;TE_count_unique&quot;: [&quot;GCC/7.3.0-2.30&quot;, &quot;OpenMPI/3.1.1&quot;, &quot;Subread/1.6.3&quot;],
        &quot;normalize_TE_counts&quot;:[&quot;GCC/9.3.0&quot;, &quot;OpenMPI/4.0.3&quot;, &quot;Seurat/3.1.5-R-4.0.0&quot;],
        &quot;plot_TE_expression&quot;:[&quot;GCC/9.3.0&quot;, &quot;OpenMPI/4.0.3&quot;, &quot;Seurat/3.1.5-R-4.0.0&quot;]
}
</code></pre>
<p>If you don't need to load any modules for a software to be available, you can leave the respective list empty.</p>
<hr />
<p>You can read the functions' documentation and some tutorials at https://ra7555ga-s.github.io/truster/</p>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="workflow/" class="btn btn-neutral float-right" title="Workflow">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
    
    
      <span><a href="workflow/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script>var base_url = '.';</script>
    <script src="js/theme_extra.js" defer></script>
    <script src="js/theme.js" defer></script>
      <script src="search/main.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>

<!--
MkDocs version : 1.4.2
Build Date UTC : 2023-01-30 11:47:48.366175+00:00
-->
